### DISCUSSION TITLE
Can I obtain a simil file like the file from the command --save-txt of Yolov5?

### DISCUSSION THREAD

**Discussion by claudio9russo7** [obss/sahi#662] [Q&A] [answered]
Hello
i didn't find anything but maybe i'm looking wrong, is it possible to get a txt file like the one you get by doing the inference with yolov5 on the pictures?

**Comment by fcakyon** [comment]
@claudio9russo7 you can export coco result json file by providing dataset json to the predict command: https://github.com/obss/sahi/issues/356

**Comment by claudio9russo7** [comment]
@fcakyon Ok, thank you. This is true only if they are already images with annotation or not?

**Comment by claudio9russo7** [comment]
I found similar the --export_pickle command.

**Comment by fcakyon** [comment, marked as answer]
Yes, that arg is relevant 👍 @claudio9russo7 does `--export_pickle` solve your issue?

**Comment by claudio9russo7** [comment]
Yep, solve my issue, I just want to suggest adding a possible choice of export format like .json or .txt :D

**Comment by fcakyon** [comment]
@claudio9russo7 noted it, will look into that feature when I find some time 🚀

**Reply by ErToBar** [reply]
Dear Fatih,  
Did you had time to look into this yet? Using the yolo .txt format would make SAHI so much more user friendly to work with. 
Atm export_pickle=True seems to be the only option in a batch processing using predict().
I guess several people have this issue and probably everyone is writing their own code to do that job.

**Reply by ErToBar** [reply]
here is my version:;
import os
import pickle
import re
from PIL import Image

# Function to extract details from object prediction
def extract_details_from_prediction(obj_pred):
    bbox_values = re.search(r'<\((.*?)\)', str(obj_pred.bbox)).group(1)
    bbox = list(map(float, bbox_values.split(',')))

    score = float(re.search(r'value: (.*?)[,>]', str(obj_pred.score)).group(1))

    category_values = re.search(r'id: (.*?),', str(obj_pred.category)).group(1)
    category = int(category_values.strip())

    return bbox, score, category

# Path of the visuals directory
visuals_path = ""

# Path of the pickles directory
pickles_path = ""

# Output path
output_path = ""

# Iterate over all images in the visuals directory
for image_file_name in os.listdir(visuals_path):
    image_path = os.path.join(visuals_path, image_file_name)

    # Corresponding pickle file name and path
    pickle_file_name = os.path.splitext(image_file_name)[0] + ".pickle"
    pickle_file_path = os.path.join(pickles_path, pickle_file_name)

    # Check if the pickle file exists
    if not os.path.exists(pickle_file_path):
        print(f"No corresponding pickle file found for image: {image_file_name}")
        continue

    # Load the pickle file data
    with open(pickle_file_path, 'rb') as f:
        data = pickle.load(f)

    # Get the dimensions of the image
    image = Image.open(image_path)
    image_width, image_height = image.size

    # Iterate over predictions and write to file
    for obj_pred in data:
        # Extract details from prediction
        bbox, score, category = extract_details_from_prediction(obj_pred)

        # Calculate the center of the box, width and height
        bbox_center_x = (bbox[0] + bbox[2]) / 2.0
        bbox_center_y = (bbox[1] + bbox[3]) / 2.0
        bbox_width = bbox[2] - bbox[0]
        bbox_height = bbox[3] - bbox[1]

        # Normalize the values
        bbox_center_x /= image_width
        bbox_center_y /= image_height
        bbox_width /= image_width
        bbox_height /= image_height

        # Save to the output file
        output_file_name = os.path.splitext(image_file_name)[0] + ".txt"
        output_file_path = os.path.join(output_path, output_file_name)

        with open(output_file_path, "a") as out_f:  # use "a" to append to file for multiple predictions
            out_f.write(f"{category} {bbox_center_x} {bbox_center_y} {bbox_width} {bbox_height} {score}\n")

**Reply by brianbattaile** [reply]
As a beginner, I tried to implement the above code but it threw all kinds of errors.  What exactly should this code look like as a whole unit and at what points does a user need to fill in paths etc.?  I get I need to add a visuals_path, pickles_path and output_path.  What else am I missing?  I tried this without success

import os
import pickle
import re
from PIL import Image

def extract_details_from_prediction(obj_pred):
    bbox_values = re.search(r'<((.*?))', str(obj_pred.bbox)).group(1)
    bbox = list(map(float, bbox_values.split(',')))

    core = float(re.search(r'value: (.*?)[,>]', str(obj_pred.score)).group(1))

    category_values = re.search(r'id: (.*?),', str(obj_pred.category)).group(1)
    category = int(category_values.strip())

    return bbox, score, category

visuals_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp\visuals"
pickles_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp\pickles"
output_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp"

for image_file_name in os.listdir(visuals_path):
    image_path = os.path.join(visuals_path, image_file_name)

    # Corresponding pickle file name and path
    pickle_file_name = os.path.splitext(image_file_name)[0] + ".pickle"
    pickle_file_path = os.path.join(pickles_path, pickle_file_name)

    # Check if the pickle file exists
    if not os.path.exists(pickle_file_path):
        print(f"No corresponding pickle file found for image: {image_file_name}")
        continue

    # Load the pickle file data
    with open(pickle_file_path, 'rb') as f:
        data = pickle.load(f)

    # Get the dimensions of the image
    image = Image.open(image_path)
    image_width, image_height = image.size

    # Iterate over predictions and write to file
    for obj_pred in data:
        # Extract details from prediction
        bbox, score, category = extract_details_from_prediction(obj_pred)

        # Calculate the center of the box, width and height
        bbox_center_x = (bbox[0] + bbox[2]) / 2.0
        bbox_center_y = (bbox[1] + bbox[3]) / 2.0
        bbox_width = bbox[2] - bbox[0]
        bbox_height = bbox[3] - bbox[1]

        # Normalize the values
        bbox_center_x /= image_width
        bbox_center_y /= image_height
        bbox_width /= image_width
        bbox_height /= image_height

        # Save to the output file
        output_file_name = os.path.splitext(image_file_name)[0] + ".txt"
        output_file_path = os.path.join(output_path, output_file_name)

        with open(output_file_path, "a") as out_f:  # use "a" to append to file for multiple predictions
            out_f.write(f"{category} {bbox_center_x} {bbox_center_y} {bbox_width} {bbox_height} {score}\n")

**Reply by brianbattaile** [reply]
Dang, I have no idea why the code section formatted the way it did in the previous post...let me try again.

import os
import pickle
import re
from PIL import Image

def extract_details_from_prediction(obj_pred):
    bbox_values = re.search(r'<((.*?))', str(obj_pred.bbox)).group(1)
    bbox = list(map(float, bbox_values.split(',')))

    core = float(re.search(r'value: (.*?)[,>]', str(obj_pred.score)).group(1))

    category_values = re.search(r'id: (.*?),', str(obj_pred.category)).group(1)
    category = int(category_values.strip())

    return bbox, score, category

visuals_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp\visuals"
pickles_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp\pickles"
output_path = r"C:\Users\Green Sturgeon\AI_Project\Test\runs\predict\exp"

for image_file_name in os.listdir(visuals_path):
    image_path = os.path.join(visuals_path, image_file_name)

    # Corresponding pickle file name and path
    pickle_file_name = os.path.splitext(image_file_name)[0] + ".pickle"
    pickle_file_path = os.path.join(pickles_path, pickle_file_name)

    # Check if the pickle file exists
    if not os.path.exists(pickle_file_path):
        print(f"No corresponding pickle file found for image: {image_file_name}")
        continue

    # Load the pickle file data
    with open(pickle_file_path, 'rb') as f:
        data = pickle.load(f)

    # Get the dimensions of the image
    image = Image.open(image_path)
    image_width, image_height = image.size

    # Iterate over predictions and write to file
    for obj_pred in data:
        # Extract details from prediction
        bbox, score, category = extract_details_from_prediction(obj_pred)

        # Calculate the center of the box, width and height
        bbox_center_x = (bbox[0] + bbox[2]) / 2.0
        bbox_center_y = (bbox[1] + bbox[3]) / 2.0
        bbox_width = bbox[2] - bbox[0]
        bbox_height = bbox[3] - bbox[1]

        # Normalize the values
        bbox_center_x /= image_width
        bbox_center_y /= image_height
        bbox_width /= image_width
        bbox_height /= image_height

        # Save to the output file
        output_file_name = os.path.splitext(image_file_name)[0] + ".txt"
        output_file_path = os.path.join(output_path, output_file_name)

        with open(output_file_path, "a") as out_f:  # use "a" to append to file for multiple predictions
            out_f.write(f"{category} {bbox_center_x} {bbox_center_y} {bbox_width} {bbox_height} {score}\n")

**Reply by brianbattaile** [reply]
Still wrong, I'm clueless.

**Reply by andressrodrl** [reply]
https://github.com/andresinsitu/sahi_custom/tree/export_txt

Ive modified sahi so you can save the predictions txt in yolo format. You can use it like this:

Add the argument ``export_yolo=True`` or the argument ``export_yolo_conf=True`` to the predict function. The first one only saves xywh and the second one also saves confidence so xywhc . Results will be saved to a "labels" folder

**Comment by brianbattaile** [comment]
How can I install your version into my local python environment so I can
use it instead of the standard SAHI version I already have installed there?
